src/krylov/krylov.h

```diff
@@ -1109,11 +1152,8 @@ typedef struct
 
 typedef struct
 {
-   HYPRE_Real   tol;
-   HYPRE_Real   atolf;
-   HYPRE_Real   cf_tol;
-   HYPRE_Real   a_tol;
-   HYPRE_Real   rtol;
+   void        *conv_params;
```

Comment:
Why does this have to be `void *`?  I think we should be able to make this opaque to this header file while still retaining some type checking by doing something similar to `hypre_SStructGrid_struct` in `sstruct_mv/HYPRE_sstruct_mv.h` for example.  In that header file, `HYPRE_SStructGrid` is used in many of the function prototypes, but it's definition is opaque in that header file.  We give it a specific definition in the C source files by way of the lower-case header file.  It looks like you need to know the type of the `void *` once you get into the `pcg.c` file.

Reply:
@rfalgout indeed I can do something similar for this. The type of the struct will be inferred by the precision in which we are building functions in pcg.c. So the struct's precision type will be the same as the function's.